pkg/config/system_mask.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package config
+
+import (
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+)
+
+// SystemConfigMask is a mask that can be applied to a set of system config
+// entries to filter out unwanted entries.
+type SystemConfigMask struct {
+	keys []roachpb.Key
+}
+
+// MakeSystemConfigMask constructs a new SystemConfigMask that passes through
+// only the specified keys when applied.
+func MakeSystemConfigMask(allowed ...roachpb.Key) SystemConfigMask {
+	sort.Slice(allowed, func(i, j int) bool {
+		return allowed[i].Compare(allowed[j]) < 0
+	})
+	return SystemConfigMask{keys: allowed}
+}
+
+// Apply applies the mask to the provided set of system config entries,
+// returning a filtered down set of entries.
+func (m SystemConfigMask) Apply(entries SystemConfigEntries) SystemConfigEntries {
+	var res SystemConfigEntries
+	for _, key := range m.keys {
+		i := sort.Search(len(entries.Values), func(i int) bool {
```

Comment:
if `entries.Values` is also sorted (as it is?), shouldn't it be possible to "merge search" here by discarding `entries.Values[:i+1]` after each hit? Probably not worth it though, given `m.keys` is small in practice.

Reply:
Yes, this seems possible, but it's probably not worth the complexity. I actually thought of doing so while writing this, and then remembered that at some point during the TLA+ training, you convinced me with some fancy math that doing this was still `O(m*log(n))`.